source/simulator/core.cc

```diff
@@ -1426,10 +1426,11 @@ namespace aspect
               build_advection_preconditioner(AdvectionField::composition(c),
                                              C_preconditioner);
               solve_advection(AdvectionField::composition(c));
-              current_linearization_point.block(introspection.block_indices.compositional_fields[c])
-                = solution.block(introspection.block_indices.compositional_fields[c]);
             }
 
+          for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+            current_linearization_point.block(introspection.block_indices.compositional_fields[c])
+              = solution.block(introspection.block_indices.compositional_fields[c]);
```

Comment:
Before this change we used the old_solution in the assembly of the composition advection, so composition 2 did not get to see the updated values for composition 1, it got the values for composition 1 from the time step before (so in principle, for the IMPES solver scheme nothing should have changed, except that we use the extrapolated values from the current linearization point instead of the old solution now). 
The only change I want to make here is that both fields get to see the values that are updated during nonlinear iterations in case the fields depend on each other (or on other solution variables). 

We can certainly think about schemes where compositional fields already get to see updated values of fields that are solved previously in the same time step, but then maybe we would have to implement this as a new scheme.